### PR TITLE
[ValSet-Pref] Add Msgs for Delegate and Undelegate Tokens

### PR DESCRIPTION
**Pull request by stackman27**
Part of : https://github.com/osmosis-labs/osmosis/issues/2579

## What is the purpose of the change

This message `MsgDelegateToValidatorSet` allows delegators to delegate to a set of defined validator set.  For ex: delegate 10osmo with validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2} our delegate logic would attempt to delegate 5osmo to A , 2osmo to B, 3osmo to C 

## Brief Changelog
n/a

## Testing and Verifying

Added tests for TestDelegateToValidatorSet

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/valset-pref/validator_set.go
Comment (stackman27) [inline x/valset-pref/validator_set.go:58]: Any suggestion on how i can retrieve this multiplier in a better way? I know in superfluid we store the existing multiplier (https://github.com/osmosis-labs/osmosis/blob/main/x/superfluid/keeper/twap_price.go#L36)

Comment (stackman27) [reply x/valset-pref/validator_set.go]: Resolved! No need to multiply manually, expect everything to be in smallest denom

**Comment by faddat** [comment]
why?

not that I have any issue with it, it seems cool,  but curious on thinking here

**Comment by stackman27** [comment]
> why?
> 
> not that I have any issue with it, it seems cool, but curious on thinking here

Hi @faddat! This feature can technically be done off-chain today with front-ends, having preference lists stored locally, but Having this on chain:
1. eases front-end code burden that is already very high in cosmos
2. reduces front-end lock-in, giving you the same UX across multiple front-ends
3. Opens door for features such as auto-staking

File: x/valset-pref/msg_server_test.go
Comment (czarcas7ic) [inline x/valset-pref/msg_server_test.go:143]:
```suggestion
			name:           "Delegate to valid validators",
```

File: x/valset-pref/msg_server_test.go
Comment (czarcas7ic) [inline x/valset-pref/msg_server_test.go:165]:
```suggestion
			name:       "User does not have enough tokens to stake",
```

File: x/valset-pref/msg_server_test.go
Comment (czarcas7ic) [inline x/valset-pref/msg_server_test.go:146]: Just fyi this abstraction made me read it twice before I got what was happening. Maybe in the comments just explicitly state in the comments that test's val set is defined as 50%,30%,and 20% instead of inferring it :)

File: x/valset-pref/msg_server_test.go
Comment (czarcas7ic) [inline x/valset-pref/msg_server_test.go:195]:
```suggestion
					// guarantees that the delegator exists because we check it in DelegateToValidatorSet
```

File: x/valset-pref/validator_set.go
Comment (czarcas7ic) [inline x/valset-pref/validator_set.go:32]: Mmm, idk if this TODO is worth leaving in the code base. Discussion should happen first with community before considering any implementation

File: x/valset-pref/validator_set.go
Comment (czarcas7ic) [inline x/valset-pref/validator_set.go:65]: The only issue I see with this is if delegating to a tombstoned validator produces an error. If this is the case, If we are halfway through the loop and we hit a tombstoned validator, it seems like we would error and the remainder of the delegation doesn't get delegated. There should be some kind of contract here that we either delegate all or none. Should check if delegating to tombstoned validators produces error. If it does, we need to check if its tombstoned prior to this. If not, can probably leave as is.

Comment (stackman27) [reply x/valset-pref/validator_set.go:65]: hmm one option would be to loop through the val-set before delegation and check if there are any jailed / invalid validators or not using `[IsValidatorJailed](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/validator.go#L480)`  or [GetValidator](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/validator.go#L15) 

And if there is any jailed validator we DONOT delegate anything. lmk what you think

File: x/valset-pref/msg_server_test.go
Comment (mattverse) [inline x/valset-pref/msg_server_test.go:197]: Can we add a check on user balance and confirmed that it has been reduced as well?

File: x/valset-pref/msg_server_test.go
Comment (mattverse) [inline x/valset-pref/msg_server_test.go:141]: would it be possible to add a test case that deals delegating coins that wouldn't be calculated perfectly under the ratio and test the rounding conditions? Ex) 739 or similar

File: x/valset-pref/msg_server_test.go
Comment (mattverse) [inline x/valset-pref/msg_server_test.go:142]: Ditto with the comment above regarding jailed validator. Consider adding a test case on it as well

File: x/valset-pref/validator_set.go
Comment (stackman27) [reply x/valset-pref/validator_set.go:65]: @mattverse we talked about this in the chain dev call. If we loop through the validators and check their jail, tombstoned &  unbonding status it might be very gas intensive, therefore i believe we can just have simple validator check like `GetValidator` and leave this as is. 

Another reason why we are being lenient with this check is that if in any iteration of the loop the validator is faulty, then the function returns an error and the entire message is reverted. **We are doing many iterations of `Delegate` function using only 1 message so we might have a leeway here**

Comment (mattverse) [reply x/valset-pref/validator_set.go:65]: Just to confirm, that means that we're still adding a check to check if a validator is jailed, and if so, reverts the entire message?

File: x/valset-pref/validator_set.go
Comment (stackman27) [reply x/valset-pref/validator_set.go:57]: aah gotchu will do

File: x/valset-pref/validator_set.go
Comment (stackman27) [reply x/valset-pref/validator_set.go:65]: we have `GetValidator` that runs in ever iteration of `Val-Set` which checks that

File: x/valset-pref/msg_server_test.go
Comment (stackman27) [reply x/valset-pref/msg_server_test.go:142]: hmm not sure how i can create a test scenario where a validator is jailed and user tries to delegate to that validator  🤔

File: x/valset-pref/msg_server_test.go
Comment (stackman27) [reply x/valset-pref/msg_server_test.go:143]: added it here [`45964da`](https://github.com/osmosis-labs/osmosis/pull/3260/commits/45964da70bb84d553d3cc72ce3a796411b58217f)

File: x/valset-pref/msg_server_test.go
Comment (stackman27) [reply x/valset-pref/msg_server_test.go:196]: added [here](https://github.com/osmosis-labs/osmosis/pull/3260/commits/09ee056004a254b10660d332b42cdbce1d2e047a#)

**Comment by stackman27** [comment]
Thank you @mattverse and @czarcas7ic for the feedback. Everything should be resolved now :))

File: x/valset-pref/msg_server_test.go
Comment (stackman27) [reply x/valset-pref/msg_server_test.go:144]: So i did some digging around and found out that "It is possible to delegate to a jailed validator, the only difference being it will not be added to the power index until it is unjailed" therefore, i dont think we can add a very indepth test

File: x/valset-pref/msg_server_test.go
Comment (czarcas7ic) [inline x/valset-pref/msg_server_test.go:140]:
```suggestion
		expectedShares []sdk.Dec // expected shares after delegation
```

File: x/valset-pref/msg_server_test.go
Comment (czarcas7ic) [inline x/valset-pref/msg_server_test.go:187]: why not check the exact amount rather than inequalities?

File: x/valset-pref/msg_server_test.go
Comment (czarcas7ic) [inline x/valset-pref/msg_server_test.go:194]:
```suggestion
					// guarantees that the delegator exists because we check it in DelegateToValidatorSet
```

File: x/valset-pref/msg_server_test.go
Comment (czarcas7ic) [inline x/valset-pref/msg_server_test.go:215]: since its only one test case here you can just say "unstake half from a"

File: x/valset-pref/validator_set.go
Comment (czarcas7ic) [inline x/valset-pref/validator_set.go:57]: I think the only one we still are curious about is tombstoned

**Review by czarcas7ic — APPROVED** [review]
This LGTM, well done! Once we get another review here and this gets merged, please ensure to test this on a localosmosis instance with all the applicable CLI commands you used. This way we can easily test come next upgrade!

File: x/valset-pref/msg_server_test.go
Comment (stackman27) [reply x/valset-pref/msg_server_test.go:187]: Yup will do

**Comment by stackman27** [comment]
Thank you @czarcas7ic addressed all you comments and added more test coverage [`06f64d`](https://github.com/osmosis-labs/osmosis/commit/06f64dc0f7e410f1d7a9fdd6d037defdd787de79)

still looking on tombstoned validators and ways to approach them

**Comment by czarcas7ic** [comment]
@stackman27 that comment was more to add it to the TODO, and not a blocker for this PR IMO

File: x/twap/api.go
Comment (mattverse) [inline x/twap/api.go:96]: doesnt this change to `getTwapToNow` anyways if `endTime.Equal(ctx.BlockTime())`?

File: x/valset-pref/msg_server_test.go
Comment (mattverse) [reply x/valset-pref/msg_server_test.go:144]: how about unbonded or unbonding validators?

File: x/valset-pref/validator_set.go
Comment (mattverse) [reply x/valset-pref/validator_set.go:57]: TODO: change comments

File: x/valset-pref/validator_set.go
Comment (mattverse) [inline x/valset-pref/validator_set.go:93]:
```suggestion
		return fmt.Errorf("The undelegate total do not add up with the amount calculated from weights expected %s got %s", tokenAmt, totalAmountFromWeights)
```

File: x/valset-pref/msg_server_test.go
Comment (mattverse) [reply x/valset-pref/msg_server_test.go:143]: cannot see it due to force push :( , can you point me to the specific test case?

**Review by mattverse — APPROVED** [review]
LGTM! 

Left some minor comments I would love to see before merging!

File: x/twap/api.go
Comment (stackman27) [reply x/twap/api.go:96]: oops this change wasn't supposed to be in this PR, but endTime = BlockTime logic is covered in line 84. We can follow up this conversation in this PR https://github.com/osmosis-labs/osmosis/pull/3529

File: x/valset-pref/msg_server_test.go
Comment (stackman27) [reply x/valset-pref/msg_server_test.go:143]: Yup, i changed the set-validator weights to have upto 3 decimal places https://github.com/osmosis-labs/osmosis/pull/3260/files#diff-102acff6e2fd0c8c3b2b95afe0c166f91100a1dcbc1204c40a0ff981823504a3R31

File: x/valset-pref/msg_server_test.go
Comment (stackman27) [reply x/valset-pref/msg_server_test.go:144]: Here's what i found; 
1. unbonded: Validators will be removed from active set but **still can receive delegations**
2. unbonding: Validators leaves active set either through slashing, jailing etc, **all delegations can be unbonded but will have to wait for unbonding time to exhaust** 

I am hoping to test all these in either e2e or testnet

**Comment by ValarDragon** [comment]
Merging due to approvals, and this PR not being merged blocking things

**Pull request merged**
